### DISCUSSION TITLE
Section 4: Remix Fund Me - Problem: Anybody, not just the owner can withdraw funds

### DISCUSSION THREAD

**Discussion by nox-dave** [Cyfrin/foundry-full-course-cu#1967] [Q&A]
Greetings everybody. 

So I have been working through FundMe.sol project by following along the 'Learn Solidity, Blockchain Development, & Smart Contracts | Powered By AI - Full Course (0 - 6)' video on Youtube.

In the video 'Learn Solidity, Blockchain Development, & Smart Contracts | Powered By AI - Full Course (0 - 6)'  at the 5:33:00 mark Patrick demonstrates how only the owner of the contract (Account 1 in this case) could withdraw funds. When he tried to withdraw funds with Account 2 it did not work.

The problem:  After I have deployed my contract and deposited a fund any account is able to withdraw the funds, and not just the owner of the contract. My code more or less looks very much the same as those in the repository: https://github.com/Cyfrin/remix-fund-me-cu
Actually, what I decided to do was to use the exact same code as that in the repository, and the problem was still there.

This problem really bothers me and I had to hear if anybody else knew what was going on.
Thanks

**Comment by EngrPips** [comment]
Please show your code here so we can help look through it and catch what is wrong. Sometimes, just sharing the portion of the code that is not working properly is enough to help you catch the bug. Also, sharing the code here helps others gain more context and learn from the discussion.

**Comment by nox-dave** [comment]
FundMe.sol
```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.19;

import {PriceConverter} from "./PriceConverter.sol"; 


error NotOwner();

contract FundMe {


    using PriceConverter for uint256;   

    uint256 public constant MINIMUM_USD = 5e18; 
                                        

        address[] public funders; 
        mapping(address funder => uint256 amountFunded) public addressToAmountFunded; 


            address public immutable i_owner; 

            constructor() {   
                i_owner= msg.sender;
            }       


            function fund() public payable {   
                require(msg.value.getConversionRate() >= MINIMUM_USD, "didn't send enough");  
                funders.push(msg.sender); 
                addressToAmountFunded[msg.sender] = addressToAmountFunded[msg.sender] + msg.value; 
            }

            modifier onlyOwner() { 
            //    require(msg.sender == i_owner, "Sender is not the owner!");
                if(msg.sender != i_owner) { revert NotOwner(); }
                _;
            }

            function withdraw() public onlyOwner {   
                for(uint256 funderIndex = 0; funderIndex < funders.length; funderIndex++){ 
                    address funder = funders[funderIndex]; 
                    addressToAmountFunded[funder] = 0;  
                }


            funders = new address[](0);

            // Transfer
        //  payable(msg.sender).transfer(address(this).balance);

            // Send
        //  bool sendSuccess = payable(msg.sender).send(address(this).balance);
        //  require(sendSuccess, "Send failed");

            // Call
            (bool callSuccess, ) = payable(msg.sender).call{value: address(this).balance}("");
            require(callSuccess, "Call failed");
        }

        receive() external payable { 
            fund();
        }       

        fallback() external payable {
            fund();
        }

}

```

PriceConverter.sol
```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.19;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/shared/interfaces/AggregatorV3Interface.sol"; 

library PriceConverter {


        function getPrice() internal view returns(uint256){
          //  Address : 0x694AA1769357215DE4FAC081bf1f309aDC325306       
          //  ABI : Check

            AggregatorV3Interface dataFeed = AggregatorV3Interface(0x694AA1769357215DE4FAC081bf1f309aDC325306); 
            (,int256 price,,,) = dataFeed.latestRoundData();  
            return uint256(price * 1e10);                        
        }


        function getConversionRate(uint256 ethAmount) internal view returns(uint256){ 
            uint256 ethPrice = getPrice(); 
            uint256 ethAmountInUsd = (ethPrice * ethAmount) / 1e18;
            return ethAmountInUsd;
        }  

        function getVersion() internal view returns (uint256) {  
            return AggregatorV3Interface(0x694AA1769357215DE4FAC081bf1f309aDC325306).version();
        }

}
```

**Comment by EngrPips** [comment]
Honestly, I can't see anything wrong here. @rasta-dave, please run your code in remix. Call the `fund` function on your code in remix and show the remix output.

**Comment by nox-dave** [comment]
I set my environment as 'Injected Provider - Metamask'. 
I deploy the contract from my Account 1 and I press 'fund'.  What is returned is: 

Gas estimation failed
Gas estimation errored with the following message (see below). The transaction execution will likely fail. Do you want to force sending?
Error happened while trying to execute a function inside a smart contract
Eip838ExecutionError: execution reverted: didn't send enough

In the 'value' field I fill in: 10000000000000000 wei.

I press 'fund' again and the deposit was successful.

When I look at 'i_owner' it indeed shows the correct address of my Account 1.

I press 'withdraw' (as Account 1) and Metamask opens up to confirm the withdraw (to Account 1).
I then change to Account 2 and press 'withdraw' again and instead of having an error.. Metamask opens up to confirm the withdraw. 

That is what makes no sense to me since Account 2 should be denied to withdraw. No error appears.

Likewise, I have copy-pasted the code from https://github.com/Cyfrin/remix-fund-me-cu and I STILL had the same problem which should not happen. This makes me wonder if something is wrong with my Remix or Metamask on my side since nothing seems to be wrong with the code.

**Reply by EngrPips** [reply]
Honestly, it is hard for me to tell what is wrong. With the code you shared, I could not fund the contract at all talk less of withdrawing. Also, I will share my code so you can try it with mine. Let's see if you still have the same problem.

**FundMe.sol**

```js
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.23;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";
contract FundMe {

    using PriceConverter for uint256;

    uint256 constant MINIMUM_ALLOWED_VALUE = 5e18;
    address immutable  PRICE_FEED_ADDRESS;
    address payable immutable  i_owner; 

    address[] private s_funders;
    mapping(address funder => uint256 amount) private s_addressToAmountFunded;

    constructor(address  _priceFeedAddress) {
        PRICE_FEED_ADDRESS = _priceFeedAddress;
        i_owner = payable(msg.sender);
    }

    function fund() public payable {
        uint256 receivedValueInUSD = msg.value.getConversionRate(PRICE_FEED_ADDRESS);
        require(receivedValueInUSD >= MINIMUM_ALLOWED_VALUE, "You didn't send enough ether");
        s_funders.push(msg.sender);
        s_addressToAmountFunded[msg.sender] += msg.value; 
    }

    function withdraw() public {
        require(msg.sender == i_owner, "You are not the owner of the contract");
        uint256 fundersLength = s_funders.length;
        for(uint256 index; index < fundersLength; index++){
            address currentAddress = s_funders[index];
            s_addressToAmountFunded[currentAddress] = 0;
        }
        s_funders = new address[](0);
        (bool callSuccess, ) = i_owner.call{value: address(this).balance}("");
        require(callSuccess == true, "Sending ether to owner failed");
    }

    function getOwner() external view returns(address) {
        return i_owner;
    }

    function getPriceFeedVersion() external view returns(uint256) {
        return AggregatorV3Interface(PRICE_FEED_ADDRESS).version();
    }

    function getAddressToAmountFunded(address _address) external view returns(uint256) {
        return s_addressToAmountFunded[_address];
    }

    function getAddressAtIndex(uint256 _index) external view returns(address) {
        return s_funders[_index];
    }

    receive() external payable  {
        fund();
    }

    fallback() external payable { 
        fund();
    }
}
```

**PriceConverter.sol**

```js
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.23;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";

library PriceConverter {
    // ETHUSD PRICEFEED ON SEPOLIA => 0x694AA1769357215DE4FAC081bf1f309aDC325306

    function getConversionRate(uint256 _ethValueToConvert, address _priceFeedAddress) internal view returns(uint256) {
        uint256 price = getPrice(_priceFeedAddress);
        uint256 receivedValueInUSD = _ethValueToConvert * (price * 10 ** 10);
        return receivedValueInUSD;
    }

    function getPrice(address _priceFeedAddress) internal view returns(uint256) {
        AggregatorV3Interface priceFeedContract = AggregatorV3Interface(_priceFeedAddress);
        (,int answer,,,) = priceFeedContract.latestRoundData();
        uint256 price = uint256(answer);
        return price;
    }

    
}
```

**Comment by nox-dave** [comment]
Thank you so much for showing me your code, EngrPips.

Using your code I reproduced all of the steps:

The environment on Remix is 'Injected Provider - Metamask'
I insert the Price Fee Address (0x694AA1769357215DE4FAC081bf1f309aDC325306) and I deploy the contract using Account 1.
I then fund the contract with 10000000000000000 wei.
Using Account 1 I can successfully withdraw funds.
I switch to Account 2, and it still allows me to withdraw funds.

The problem persists on my part. There is no doubt the issue has nothing to do with the code, but it must be somewhere on my side in terms of Remix or Metamask. Where I do not know.

UPDATE:
Using Account 2 I proceeded to complete the withdrawal, and the ETH did NOT transfer to Account 2, but went into Account 1 instead. So technically the safety mechanism worked. I did NOT get any error however which is weird, but the ETH transferred automatically to the owner of the contract instead of going into Account 2.

**Comment by EngrPips** [comment]
That seems okay but honestly, `Account 2` is not supposed to be able to call withdraw at all. I went ahead and tested it on my end when I tried to withdraw with an account other than the account that created the contract, I got this error.

<img width="1440" alt="Screenshot 2024-07-09 at 15 38 26" src="https://github.com/Cyfrin/foundry-full-course-cu/assets/95760918/b9eaff8f-20c1-40a1-9ea3-4ad20f3504dd">

But when I called `withdraw` with the account that created the contract, then it went through. That should be the normal behavior.

<img width="1440" alt="Screenshot 2024-07-09 at 15 39 30" src="https://github.com/Cyfrin/foundry-full-course-cu/assets/95760918/2ca47c5d-5b8e-485b-a6a2-e62aecaf05db">

**Comment by nox-dave** [comment]
Absolutely. I have no clue why I don't get the same error.

When I try to withdraw as Account 1 which is the owner of the contract
![github1](https://github.com/Cyfrin/foundry-full-course-cu/assets/127548786/ea7e441c-cc4a-496e-bd90-ab3abdc580b9)


When I try to withdraw as Account 2 (no error)
![github2](https://github.com/Cyfrin/foundry-full-course-cu/assets/127548786/e831ace2-5524-4226-b479-d33cc86bfc30)

**Reply by EngrPips** [reply]
I think you are only using `account 1` all through for the process. maybe retry the process then open the transaction output and check the sender for each of the transactions.
